osutil/disks/disks.go

```diff
@@ -173,6 +195,89 @@ func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
 	}
 
+	if opts != nil && opts.IsDecryptedDevice {
+		// verify that the mount point is indeed a mapper device, it should:
+		// 1. have DEVTYPE == disk from udev
+		// 2. have dm files in the sysfs entry for the maj:min of the device
+		if props["DEVTYPE"] != "disk" {
+			// not a decrypted device
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device: devtype is not disk (is %s)", partMountPointSource, props["DEVTYPE"])
+		}
+
+		// TODO:UC20: currently, we effectively parse the DM_UUID env variable
+		//            that is set for the mapper device volume, but doing so is
+		//            actually wrong, since the value of DM_UUID is an
+		//            implementation detail that depends on the subsystem
+		//            "owner" of the device such that the prefix is considered
+		//            the owner and the suffix is private data owned by the
+		//            subsystem. In our case, in UC20 initramfs, we have the
+		//            device "owned" by systemd-cryptsetup, so we should ideally
+		//            parse that the first part of DM_UUID matches CRYPT- and
+		//            then use `cryptsetup status` (since CRYPT indicates it is
+		//            "owned" by cryptsetup) to get more information on the
+		//            device sufficient for our purposes to find the encrypted
+		//            device/partition underneath the mapper.
+		//            However we don't currently have cryptsetup in the initrd,
+		//            so we can't do that yet :-(
+
+		// TODO:UC20: these files are also likely readable through udev env
+		//            properties, but it's unclear if reading there is reliable
+		//            or not, given that these variables have been observed to
+		//            be missing from the initrd previously, and are not
+		//            available at all during userspace on UC20 for some reason
+		dmUUID, err := ioutil.ReadFile(filepath.Join(devBlockDir, d.Dev(), "dm", "uuid"))
+		if err != nil && os.IsNotExist(err) {
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device: missing device mapper metadata: no dm-uuid", partMountPointSource)
+		}
+
+		dmName, err := ioutil.ReadFile(filepath.Join(devBlockDir, d.Dev(), "dm", "name"))
+		if err != nil && os.IsNotExist(err) {
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device: missing device mapper metadata: no dm-name", partMountPointSource)
+		}
+
+		// trim the suffix of the dm name from the dm uuid to safely match the
+		// regex - the dm uuid contains the dm name, and the dm name is user
+		// controlled, so we want to remove that and just use the luks pattern
+		// to match the device uuid
+		// we are extra safe here since the dm name could be hypothetically user
+		// controlled via an external USB disk with LVM partition names, etc.
+		dmUUIDSafe := bytes.TrimSuffix(
+			bytes.TrimSpace(dmUUID),
+			append([]byte("-"), bytes.TrimSpace(dmName)...),
+		)
+		matches := luksUUIDPatternRe.FindSubmatch(dmUUIDSafe)
```

Comment:
I don't see how this is simpler? Ideally the values we are using would even be already in the udev props

Reply:
As discussed on IRC with @pedronis, the approach may be simpler because we are looking at a fixed hierarchy rather than parsing a name which isn't necessarily stable. OTOH, it may need to handle more error cases, eg. complain when slaves count is > 1, error branches on os.Readlink(), Glob and so on.

I don't have a strong opinion in either direction. With the tweaks that @mvo5 suggested the code is fine too.

Reply:
Before I say yes or no to this, I want to crack open a debugging session in the initramfs to see what the layout there actually looks like, as we have seen before there can be differences between the initramfs layout of things and the user space layout of things. I assume it will be the same since this is sysfs and one hopes that sysfs is more stable than udev, but I've gotten a bit paranoid about this kind of thing as of late.

Reply:
to be clear I don't think we should pursue this other approach until we know a reason that makes it clearly better. it's definitely more code/code paths